src/host/cmdline.cpp

```diff
@@ -27,51 +27,6 @@
 #pragma hdrstop
 using Microsoft::Console::Interactivity::ServiceLocator;
 
-// Routine Description:
-// - This routine validates a string buffer and returns the pointers of where the strings start within the buffer.
-// Arguments:
-// - Unicode - Supplies a boolean that is TRUE if the buffer contains Unicode strings, FALSE otherwise.
-// - Buffer - Supplies the buffer to be validated.
-// - Size - Supplies the size, in bytes, of the buffer to be validated.
-// - Count - Supplies the expected number of strings in the buffer.
-// ... - Supplies a pair of arguments per expected string. The first one is the expected size, in bytes, of the string
-//       and the second one receives a pointer to where the string starts.
-// Return Value:
-// - TRUE if the buffer is valid, FALSE otherwise.
-bool IsValidStringBuffer(_In_ bool Unicode, _In_reads_bytes_(Size) PVOID Buffer, _In_ ULONG Size, _In_ ULONG Count, ...)

```

Comment:
_what on earth_? is this "split string into sized chunks"?

Do you call it, like,

```
IsValidStringBuffer(true, L"Whatever", 8, 2, 2, &a, 6, &b);
```

and get back

true, `a` points to `Wh` and b points to `atever`?

Reply:
or maybe it's 16, 4, 4, 12 instead of 8, 2, 2, 6.

Reply:
Oh, man am I glad I'm not alone with that reaction... I was _really_ stumped when I tried to initially understand this function. It's my main (other) reason of wanting to remove this.

You pass this function
```
IsValidStringBuffer(true, L"a1b23c456", 9, 2, &a, 3, &b, 4, &c);
```
and the first 3 parameters will be:
* whether it's `char` or `wchar_t`
* the input buffer pointer
* the input buffer length

Afterwards a list of pairs of parameters are expected where each pair consists of a:
* `ULONG`: the expected length of this section of the string
* `void**`: a `void*` pointer that will be filled with the address at this point in the buffer

If the sum of the `ULONG` parameters in the pairs equals the length of the whole buffer it'll return `true`.

In other words, for the example above you'll get the 3 pointers pointing at:
* `a` == "a1"
* `b` == "b23"
* `c` == "c456"

This is a rather lengthy explanation, which IMO shouldn't be need given that the problem that needs to be solved is simply:
* Problem: In the condrv payload 3 strings are stored "back to back"
* Task: Extract the 3 strings

I hope that my new code is a more direct translation of that.

Reply:
```
template <int N>
std::array<std::wstring_view, N> GetAndValidateSubstrings(wstring_view in, int... lengths) {
}
```

```
auto [one, two, three] = GetAndValidate(crap, commandNameLen, otherStupidStuff)
```

no, the thing you have done is much better.

Reply:
```c++
#include <optional>
#include <string_view>

template <typename T>
struct expand_pack_as_strings {
    using type = std::wstring_view;
};

template <typename... T>
auto WeirdSplit(std::wstring_view in, T... lengths) -> std::optional<std::tuple<typename expand_pack_as_strings<T>::type...>> {
    if(in.size() < (lengths + ...)) {
        return std::nullopt;
    }
    // well, fill the rest in!
    return {};
}

void foo() {
    auto splitResult = WeirdSplit(L"Testing", 1, 2, 3, 1);
    auto [a, b, c, d] = *splitResult;
    // fucking mental
    // a, b, c, d should be wstring_views
    // or it should return nullopt
}
```

i hate it with every ounce of my being